File: LeetCode/Reverse_Linked_List/Reverse_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) {
            return head;
        }
        
        Stack<ListNode> nodeStack = new Stack<ListNode>();
        
        ListNode runner = head;
        while(runner != null) {
            nodeStack.push(runner);
            runner = runner.next;
        }
        
        head = nodeStack.peek();
        
        while(!nodeStack.isEmpty()) {
            ListNode node = nodeStack.pop();
            if(nodeStack.isEmpty()) {
                node.next = null;
            }
            else {
                node.next = nodeStack.peek();
            }
        }
        
        return head;
    }
}

File: LeetCode/Reverse_Linked_List/Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == NULL) {
            return head;
        }
        
        stack<ListNode*> nodeStack;
        ListNode* runner = head;
        while(runner != NULL) {
            nodeStack.push(runner);
            runner = runner->next;
        }
        head = nodeStack.top();
        
        while(!nodeStack.empty()) {
            ListNode* node = nodeStack.top();
            nodeStack.pop();
            if(nodeStack.empty()) {
               node->next = NULL;
            }
            else {
                node->next = nodeStack.top();
            }
        }
        
        return head;
    }
};

